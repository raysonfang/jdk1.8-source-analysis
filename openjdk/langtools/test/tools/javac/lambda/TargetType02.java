/*
 * Copyright (c) 2011, 2013, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

/*
 * @test
 * @bug 8003280
 * @summary Add lambda tests
 *  check overload resolution and target type inference w.r.t. generic methods
 * @author  Maurizio Cimadamore
 * @compile/fail/ref=TargetType02.out -XDrawDiagnostics TargetType02.java
 */

public class TargetType02 {

    interface S1<X extends Number> {
        X m(Integer x);
    }

    interface S2<X extends String> {
        abstract X m(Integer x);
    }

    static <Z extends Number> void call1(S1<Z> s) { }

    static <Z extends String> void call2(S2<Z> s) { }

    static <Z extends Number> void call3(S1<Z> s) { }
    static <Z extends String> void call3(S2<Z> s) { }

    void test() {
        call1(i -> { toString(); return i; });
        call2(i -> { toString(); return i; });
        call3(i -> { toString(); return i; });
    }
}
