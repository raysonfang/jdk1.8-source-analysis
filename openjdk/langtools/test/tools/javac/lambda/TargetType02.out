TargetType02.java:52:14: compiler.err.prob.found.req: (compiler.misc.inferred.do.not.conform.to.upper.bounds: java.lang.Integer, java.lang.String)
TargetType02.java:53:9: compiler.err.ref.ambiguous: call3, kindname.method, <Z>call3(TargetType02.S1<Z>), TargetType02, kindname.method, <Z>call3(TargetType02.S2<Z>), TargetType02
2 errors
